app/models/redirect.rb

```diff
@@ -49,4 +50,8 @@ def noncircular_redirect
     errors.add(:target_path, "redirects to itself") if source_path == target_path
   end
 
+  def article_short_path_unique
+    errors.add(:source_path, 'is already taken') if Article.where(short_path: self.source_path).exists?
```

Comment:
Opposite of above, you'll need to strip the leading `/` on `self.source_path` to match any `Article.short_path`.